Mc labs

Quinta-5 opened the pull request.

Quinta-5 committed:
Update to Current

Quinta-5 committed:
Added Pumpkins and Melons and Expanded Cactus Functionality

For Pumpkins and Melons, they now grow an additional fruit when the faction upgrade is applied. Extra fruit generation mimics vanilla requirements, except, when activated, the upgrade will always grow an additional fruit if there is space around it and the vanilla conditions are met. The extra fruit is also grown fast enough to be compatible with redstone auto farms.

For cactus, if there are neighboring blocks around where the cactus naturally grew, the factions upgrade will now wait until the cactus is broken and replant an additional cactus in its place, which will also break, leading to an extra cactus dropped. In combination with the height increase, this update should now provide an extra cactus in all circumstances.

Quinta-5 committed:
Condensed code for Crops Upgrade

No extended functionality or additional crop compatibility was added in this update. It was simply made to condense the code.

Quinta-5 committed:
Added Shrooms Combability for Crops Upgrade

Shrooms compatibility has been added to the Crops Upgrade. The Crops Upgrade should now be compatible for every sellable crop in MCLabs, excluding chorus.

Quinta-5 committed:
Imports fix 

No functionality changes, just fixing imports

Quinta-5 committed:
Update UpgradesListener.java

Quinta-5 committed:
Updatee

Quinta-5 committed:
Update Final

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -55,7 +59,7 @@ public void init() {
     @EventHandler
     public void onDeath(EntityDeathEvent e) {
-        Entity killer = e.getEntity().getKiller();
+       Entity killer = e.getEntity().getKiller();
```
```suggestion
        Entity killer = e.getEntity().getKiller();
```

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+          }
         }
+    
+    BlockGrowEvent cropGrowEvent;
```
Move all of this (206-215) into the init() function. You shouldn't have code that's settings variables outside of functions in the middle of the class. 

I'd also recommend utilizing [lazy initialization](https://learn.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization).

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    		}
     }
+    					
+    Block blockAutoAgeable;
```
Same with this

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    	}
+    }
+      		    		
+private boolean getNeighbors() {
```
Formatting/indentation

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+			} return false; 		
+	}
+
+    private void growManual(BlockGrowEvent cropGrowEvent) {
```
Formatting/indentation

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    			if ((typeAttachedStem == blockCrop.getRelative(blockface).getType()) 
+    					&& ((((Directional) (blockCrop.getRelative(blockface)).getBlockData()).getFacing() == blockface.getOppositeFace()))) {
+    					this.blockStemAttached = blockCrop.getRelative(blockface);		
+    			} return blockStemAttached; 		
```
the return should be on the following line

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    	 blockCrop.setBlockData((BlockData)(Ageable)dataCropNew);   	 
+    	}
+    
+    private Set<Material> fruitSpawnable = EnumSet.of(Material.DIRT, Material.GRASS_BLOCK, Material.PODZOL, Material.COARSE_DIRT, Material.MYCELIUM, Material.ROOTED_DIRT, Material.MOSS_BLOCK, Material.FARMLAND);
```
move to inside a function or to top of class

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
 
+    private void growExtraFruit() {
+    	Block blockAttachedStem = getAttachedStem();
+    	for (BlockFace blockface : blockFaces)
```
add braces to this for loop, makes it easier to read

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    	this.locationCrop = cropGrowEvent.getNewState().getBlock().getLocation();
+    	if ((!manualCrops.contains(typeCropNew)) && (!autoCrops.contains(typeCropNew))) { 
+    		return;
+    		}else if (manualCrops.contains(typeCropNew)) {
```
no spaces after braces hurts me

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
         }
     }
+    
+    Plugin plugin = FactionsPlugin.getPlugin(FactionsPlugin.class);
```
move this to within a function/to the top of the class

Quinta-5 replied:
moved to top of the class

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    public void onCropGrow(BlockGrowEvent e) {
+        FLocation floc = FLocation.wrap(e.getBlock().getLocation());
+        Faction factionAtLoc = Board.getInstance().getFactionAt(floc);
+        if (!factionAtLoc.isWilderness()) {
```
This pattern of checking for not in wilderness, getting the level/chance, and generating a chance, is repeated a couple times. You could extract this into a function to reduce code duplication and make future changes have a single place to edit

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+    
+    private void growCrop(BlockGrowEvent e) {
+    	this.cropGrowEvent = e;
+    	this.blockCrop = cropGrowEvent.getNewState().getBlock();
```
If these variabels are local to this function, don't use `this` `this` makes it a variable in the scope of the entire class (which is what `this` is, this class). If the variable doesn't need to persist outside of the function, you can just declare it as `blockCrop`.

If you are using it in another function, consider passing it to that function instead.

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+
+    private void growManual(BlockGrowEvent cropGrowEvent) {
+    	cropGrowEvent.setCancelled(true);
+    	 int age =((Ageable) dataCropNew).getAge();
```
spacing issues

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+	
+    private boolean canSurvive(Block e){
+    	Material below = e.getLocation().add(0.0, -1.0, 0.0).getBlock().getType();
+    	Material origin= e.getType();
```
painful spacing

Dakotaa commented on src/main/java/com/massivecraft/factions/zcore/frame/fupgrades/UpgradesListener.java:
```diff
@@ -95,41 +99,207 @@ private void lowerSpawnerDelay(SpawnerSpawnEvent e, double multiplier) {
+                        }
+                    }
+                }
+            }return true;	
```
new line for return

Quinta-5 committed:
Removed belowMaxDensity() check 

I removed the belowMaxDensity() method and its inclusion in the spreadShroom() method. The mushroom density check is already passed when the block spreads naturally, so there is no need to include it in the upgrade, and it may cause problems.

Quinta-5 committed:
Condensed Code for Crops Upgrade Activation

Quinta-5 committed:
Update

Co-authored-by: Dakota Lalach <kotalalach@gmail.com>

Quinta-5 committed:
Style/Formatting fixes

Quinta-5 committed:
General Formatting/Style Fixes

Quinta-5 committed:
From Dak, Untested

Quinta-5 committed:
From Dak, Untested with Commit Comment

Make material sets lazy-initialized and clean up function parameters

Material sets, such as ManualCrops, are now lazy-initialized, meaning the set is now class-scoped,
and is not created until it is first needed. This means the set will only ever be created once,
rather than being created each time the function is called, which is redundant because the function
will always use the same set.

Function parameters were cleaned up in cases where a function used a variable that was needlessly
class-scoped and could rather be passed into that function by the parent function calling it.
Example: The calls to growManual, growFruit, and growAutoAgeable in growCrop() are now passed in
the block data rather than grabbing it from the class variables.
